core/src/main/kotlin/org/evomaster/core/search/gene/ObjectGene.kt

```diff
@@ -20,36 +26,183 @@ import org.slf4j.LoggerFactory
+        }
 
     /*
         In theory, it is possible to have an object with no fields...
      */
     override fun canBeChildless() = true
 
     override fun copyContent(): Gene {
-        return ObjectGene(name, fields.map(Gene::copy), refType)
+        return ObjectGene(name, fixedFields.map(Gene::copy), refType, isFixed, template, additionalFields?.map {it.copy() as PairGene<StringGene, Gene> }?.toMutableList())
+    }
+
+    override fun isLocallyValid(): Boolean {
+        return getViewOfChildren().all { it.isLocallyValid() }
+    }
+
+    override fun randomize(randomness: Randomness, tryToForceNewValue: Boolean) {
+        fixedFields.filter { it.isMutable() }
+                .forEach { it.randomize(randomness, tryToForceNewValue) }
+
+        if (!isFixed){
+            Lazy.assert {
+                template != null && additionalFields != null
+            }
+            if (additionalFields!!.isNotEmpty())
+                killChildren(additionalFields!!)
+            val num = randomness.nextInt(MAX_SIZE_ADDITIONAL_FIELDS)
+            repeat(num){
+                val added = addElement(randomness)
+                if (added != null){
+                    addChild(added)
+                }
+            }
+        }
+    }
+
+    override fun customShouldApplyShallowMutation(randomness: Randomness, selectionStrategy: SubsetGeneMutationSelectionStrategy, enableAdaptiveGeneMutation: Boolean, additionalGeneMutationInfo: AdditionalGeneMutationInfo?): Boolean {
+        if (isFixed) return false
+        return randomness.nextBoolean(PROB_MODIFY_SIZE_ADDITIONAL_FIELDS)
+    }
+
+    override fun shallowMutate(randomness: Randomness, apc: AdaptiveParameterControl, mwc: MutationWeightControl, selectionStrategy: SubsetGeneMutationSelectionStrategy, enableAdaptiveGeneMutation: Boolean, additionalGeneMutationInfo: AdditionalGeneMutationInfo?): Boolean {
+        if (isFixed)
+            throw IllegalStateException("shallowMutate is not applied to fixed ObjectGene")
+
+        Lazy.assert {
+            additionalFields != null
+            template != null
+        }
+
+        if (additionalFields!!.isEmpty() || (additionalFields!!.size < MAX_SIZE_ADDITIONAL_FIELDS && randomness.nextBoolean(0.5))){
+            val added = addElement(randomness)
+            if (added != null){
+                addChild(added)
+                return true
+            }
+
+
+            if (additionalFields!!.isEmpty()){
+                log.warn("fail to apply shallowMutate for FlexibleObject, i.e., adding or removing additional field")
+                return false
+            }
+        }
+
+        val remove = randomness.choose(additionalFields!!)
+        remove.removeThisFromItsBindingGenes()
+        killChild(remove)
+        return true
+    }
+
+    private fun addElement(randomness: Randomness) : PairGene<StringGene, FlexibleGene>?{
```

Comment:
why this is method is called `add`? isn't just returning a copy, and not adding it?